doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,262 @@
+```
+
+and in the algorithms get_ctx() function
+
+``` C
+int xxx_get_fips_approved(OSSL_PARAM params[])
+{
+    p = OSSL_PARAM_locate(params, OSSL_FIPS_PARAM_APPROVED);
+    if (p != NULL && !OSSL_PARAM_set_int(p, ossl_xxx_fips_approved()))
+        return 0;
+    return 1;
+}
+```
+
+### API’s that would be used to support this are
+
+- EVP_PKEY Keygen, Encryption, Signatures, Key Exchange, KEM
+
+``` C
+EVP_PKEY_CTX_get_params(ctx, );
+```
+
+(Note that this would mean you could not use functions that hide the ctx such as EVP_PKEY_Q_keygen()!)
+
+- Ciphers
+
+``` C
+EVP_CIPHER_CTX_get_params()
+```
+
+- Digests
+
+``` C
+EVP_MD_CTX_get_params()
+```
+
+- KDF’s
+
+``` C
+EVP_KDF_CTX_get_params()
+```
+
+- MAC’s
+
+``` C
+EVP_MAC_CTX_get_params()
+```
+
+- RAND
+
+``` C
+EVP_RAND_CTX_get_params()
+```
+
+### Backwards Compatibility
+
+Previous providers do not support this operation, so they will return not approved if they are not handled.
+
+### Alternate Solution
+
+If we had different kinds of compliance requirements (something other than FIPS) either a separate getter could be added or the getter could return a int type instead of a 0 or 1..
+(e.g 1 = fips approved, 2 = some other compliance approved)
+
+Changes Required for indicators
+-------------------------------
+
+### key size >= 112 bits
+
+There are a few places where we do not enforce key size that need to be addressed.
+
+- HMAC  Which applies to all algorithms that use HMAC also.
+- CMAC
+
+### Algorithm Transitions
+
+Should we remove these algorithms completely from the fips provider, or use indicators?
+
+- DES_EDE3_ECB.  Disallowed for encryption, allowed for legacy decryption
+- DSA.  Keygen and Signing are no longer approved, not sure if verify is still approved.
+- RSA Signing using PKCSV15.  Rsa self test for sign may need to change as the default pad_mode is PKCSV15. Check if saltlen needs a indicator. Padding mode updates required in rsa_check_padding(). Check if sha1 is allowed?
+- ECDSA B & K curves are deprecated   It looks like these are still allowed. Are the approved?
+- ED25519/ED448 is now approved.
+- X25519/X448 is not approved currently. keygen would need an indicator if we allow it?
+- RSA Key transport  Padding need to be changed to either not allow padding modes other than OEAP or use an indicator.
+- RSA - It looks like SP800-131Ar2 specifies that RSA >= 2048 is approved. Can we check with the labs what to do for the legacy verification cases since we allow >=1024. Would this now be unapproved? This would apply to rsa_kem also? rsa_keygen_pairwise_test() may need to change to do a sign/verify PCT?
+- TLS1_PRF  If we are only trying to support TLS1.2 here then we should remove the tls1.0/1.1 code from the FIPS MODULE.
+
+### Digest Checks
+
+Any algorithms that use a digest need to make sure that the CAVP certificate lists all supported FIPS digests otherwise an indicator is required.
+This applies to the following algorithms:
+
+- SSKDF
+- TLS_1_3_KDF
```

Comment:
The TLS 1.3 KDF is approved by CMVP under FIPS 140-3 IG 2.4.B, Resolution 5:
```
The TLS 1.3 key derivation function documented in Section 7.1 of RFC 8446. This is considered an
approved CVL because the underlying functions performed within the TLS 1.3 KDF map to NIST
approved standards, namely: SP 800-133rev2 (Section 6.3 Option #3), SP 800-56Crev2, and SP 800-108
```
Note that RFC 8446 Appendix B.4 only specifies 5 ciphersuites, which either use SHA-256 or SHA-384. It seems that the CAVP team interprets this as a restriction on the KDF as well, so only those are testable: https://github.com/usnistgov/ACVP/commit/1df47f365f38202eabb90cc3b8faff568ea22204